llpc/patch/llpcPatchIntrinsicSimplify.h

```diff
@@ -0,0 +1,91 @@
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  llpcPatchIntrinsicSimplify.h
+ * @brief LLPC header file: contains declaration of class Llpc::PatchIntrinsicSimplify.
+ ***********************************************************************************************************************
+ */
+#pragma once
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/IR/Intrinsics.h"
+
+#include "llpcPatch.h"
+#include "llpcPipelineState.h"
+
+namespace llvm
+{
+
+class AnalysisUsage;
+class Function;
+class FunctionPass;
+class Instruction;
+class IntrinsicInst;
+class ScalarEvolution;
+class Value;
+
+} // llvm
+
+namespace Llpc
+{
+
+class Context;
+
+// =====================================================================================================================
+// Represents the LLVM pass for intrinsic simplifications.
+class PatchIntrinsicSimplify final:
+    public llvm::FunctionPass
+{
+public:
+    explicit PatchIntrinsicSimplify();
+
+    void getAnalysisUsage(llvm::AnalysisUsage& analysisUsage) const override;
+    bool runOnFunction(llvm::Function& function) override;
+
+    // -----------------------------------------------------------------------------------------------------------------
+
+    static char ID;   // ID of this pass
+
+private:
+    LLPC_DISALLOW_COPY_AND_ASSIGN(PatchIntrinsicSimplify);
+
+    bool CanSafelyConvertTo16Bit(llvm::Value* const pValue) const;
+    llvm::Value* ConvertTo16Bit(llvm::Value* const pValue, IRBuilder<>& builder) const;
+    llvm::Value* SimplifyImage(llvm::IntrinsicInst* const pIntrinsicCall,
+                               llvm::ArrayRef<uint32_t> coordOperandIndices) const;
+    llvm::Value* SimplifyTrigonometric(llvm::IntrinsicInst* const pIntrinsicCall) const;
+    llvm::Value* Simplify(llvm::IntrinsicInst* const pIntrinsicCall) const;
+
+    // -----------------------------------------------------------------------------------------------------------------
+
+    llvm::ScalarEvolution* m_pScalarEvolution; // Scalar evolution
+    llvm::LLVMContext* m_pContext;                       // LLPC context
+    llvm::Function* m_pFunction;               // Current function on analysis
+    PipelineState*  m_pPipelineState = nullptr;
```

Comment:
You can forward-declare `PipelineState` and only include the header in the cpp file.